src/runtime_src/core/tools/common/ReportMemory.cpp

```diff
@@ -0,0 +1,447 @@
+        ptMem.put("extended_info.ecc.error.uncorrectable.count", ue_cnt);
+        ptMem.put("extended_info.ecc.error.uncorrectable.first_failure_address", boost::format("0x%x") % ue_ffa);
+      }
+    }
+    std::stringstream ss(mm_buf[i]);
+    ss >> memoryUsage >> boCount;
+
+    ptMem.put("type", str);
+    ptMem.put("tag", map->m_mem_data[i].m_tag);
+    ptMem.put("enabled", map->m_mem_data[i].m_used ? true : false);
+    ptMem.put("base_address", boost::format("0x%x") % map->m_mem_data[i].m_base_address);
+    ptMem.put("range_bytes", boost::format("0x%x") % (map->m_mem_data[i].m_size << 10));
+    if (!temp_buf.empty() && temp[i] != XCL_INVALID_SENSOR_VAL)
+      ptMem.put("extended_info.temperature_C", temp[i]);
+    ptMem.put("extended_info.usage.allocated_bytes", memoryUsage);
+    ptMem.put("extended_info.usage.buffer_objects_count", boCount);
+    ptMem_array.push_back(std::make_pair("",ptMem));
+  }
+  pt.add_child(std::string("board.memory.data_streams"), ptStream_array);
+  pt.add_child(std::string("board.memory.memories"), ptMem_array );
+
+  try {
+    mm_buf = xrt_core::device_query<qr::memstat_raw>(device);
+    gbuf = xrt_core::device_query<qr::group_topology>(device);
+  } catch (const std::exception& ex){
+    pt.put("error_msg", ex.what());
+  }
+
+  if (gbuf.empty() || mm_buf.empty())
+    return pt;
+
+  const mem_topology *grp_map = (mem_topology *)gbuf.data();
+
+  for (int i = 0; i < grp_map->m_count; i++) {
+    if (grp_map->m_mem_data[i].m_used != 0) {
+      boost::property_tree::ptree ptGrp;
+      auto search = memtype_map.find((MEM_TYPE)grp_map->m_mem_data[i].m_type );
+      std::string str = search->second;
+      std::stringstream ss(mm_buf[i]);
+      ss >> memoryUsage >> boCount;
+      ptGrp.put("type", str);
+      ptGrp.put("tag", grp_map->m_mem_data[i].m_tag);
+      ptGrp.put("base_address", boost::format("0x%x") % map->m_mem_data[i].m_base_address);
+      ptGrp.put("range_bytes", boost::format("0x%x") % (grp_map->m_mem_data[i].m_size << 10));
+      ptGrp.put("extended_info.usage.allocated_bytes", memoryUsage);
+      ptGrp.put("extended_info.usage.buffer_objects_count", boCount);
+      ptGrp_array.push_back(std::make_pair("",ptGrp));
+    }
+  }
+  pt.add_child(std::string("board.memory.memory_groups"), ptGrp_array);
+  return pt;
+}
+
+void
+ReportMemory::getPropertyTreeInternal( const xrt_core::device * _pDevice, 
+                                              boost::property_tree::ptree &_pt) const
+{
+  // Defer to the 20202 format.  If we ever need to update JSON data, 
+  // Then update this method to do so.
+  getPropertyTree20202(_pDevice, _pt);
+}
+
+void 
+ReportMemory::getPropertyTree20202( const xrt_core::device * _pDevice, 
+                                           boost::property_tree::ptree &_pt) const
+{
+  _pt.add_child("mem_topology", populate_memtopology(_pDevice, "Memory Information"));
+}
+
+void 
+ReportMemory::writeReport( const xrt_core::device * _pDevice,

```

Comment:
This function is much too long, please refactor.  Since it is nicely divided into separate blocks these could be the functions.   Large functions are incomprehensible, small functions are self documenting.
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rf-single